### PR TITLE
arch.linux: simplify proc file parsing

### PR DESCRIPTION
**Pull request by mspncp**
- use a 'with' statement to simplify error handling
- open '/proc/net/dev' in text mode to avoid the need for decoding
- use a list comprehension to generate the result

_(related to issue #3856)_

### CODE REVIEW

**Comment by mspncp** [comment]
The codecov drop of  (-1.99%) compared to 3ce66d1 is a mystery to me.

Anyway, there is more in the file which can be simplified... I'll switch to draft mode.

**Comment by gpotter2** [comment]
Thanks for the cleanup ! You're right this function is written like garbage

You can safely ignore codecov (if clearly it's unrelated).

**Comment by mspncp** [comment]
@gpotter2 there are other places in the module where proc files are parsed and which could be cleaned up as well. That’s the reason why I switched to draft mode. Are you sure you don’t want to wait for it? Of course I could do it in a followup PR if you prefer it like that.

**Comment by gpotter2** [comment]
Sure thing ! Sorry about that. I wanted all tests to run. Will wait

**Comment by mspncp** [comment]
Done
 
    arch.linux: simplify proc file parsing
    
    - use a 'with' statement to simplify error handling
    - open proc files in text mode to avoid the need for decoding
    - use list comprehensions and tuple unpacking

**Comment by mspncp** [comment]
Remark: I removed the `_b` suffixes from all variable names which changed from `bytes` to `str`. To fix mypy errors and to make the naming more consistent, I had to add some `_int` suffixes to some variables in other places.

**Comment by mspncp** [comment]
Now mypy and flake8 are happy, ready for review.

**Comment by mspncp** [comment]
Last [force-push](https://github.com/secdev/scapy/compare/9f48f3051cde0da3c73d174f76e26d315072dcf3..4033577747bac18874e6af41c9bdea7bcfa45db1) addressed the [Codacy complaints](https://app.codacy.com/gh/secdev/scapy/pullRequest?prid=10970983).

File: scapy/arch/linux.py [hunk 1]
```diff
@@ -132,7 +133,8 @@ def _get_if_list():
     """
     try:
         with open("/proc/net/dev", "r") as f:
-            return [line.split(':', 1)[0].strip() for line in f.readlines()[2:]]
+            return [line.split(':', 1)[0].strip()
+                    for line in itertools.islice(f, 2, None)]
```
Comment (mspncp) [inline scapy/arch/linux.py:137]: What is the benefit of using `itertools.islice` vs. using a regular slice?

File: scapy/arch/linux.py [hunk 2]
```diff
@@ -228,66 +230,89 @@ def get_alias_address(iface_name,  # type: str
 
 def read_routes():
     # type: () -> List[Tuple[int, int, str, str, str, int]]
+    """
+    Read routes from /proc/net/route
+    """
     try:
-        with open("/proc/net/route", "r") as f:
-            entries = [line.split() for line in f]
+        with open("/proc/net/route", "r", errors='replace') as f:
```
Comment (mspncp) [inline scapy/arch/linux.py:237]: Interesting, I was aware of the fact that switching from binary mode + `plain_str()` conversion to text mode could potentially introduce decoding errors, but then I thought that (if at all) the interface names would be utf-8.

Using `errors='replace'` is an elegant solution, I didn't know that existed.

**Review by mspncp — COMMENTED** [review]
Just two questions, for my education:

File: scapy/arch/linux.py (hunk 2 above)
Comment (mspncp) [reply scapy/arch/linux.py:237]: The question here actually was: is actually possible that /proc/net/route contains non-ascii characters?

File: scapy/arch/linux.py (hunk 1 above)
Comment (mspncp) [reply scapy/arch/linux.py:137]: Ah, never mind, I checked the help myself: the main difference is that the lazy evaluation.

File: scapy/arch/linux.py (hunk 2 above)
Comment (gpotter2) [reply scapy/arch/linux.py:237]: You can hurt yourself badly for sure. (we don't support that btw, future todo)

```
user@host:~$ sudo ip link add toto type veth peer name 🥲
user@host:~$ cat /proc/net/dev
Inter-|   Receive                                                |  Transmit
 face |bytes    packets errs drop fifo frame compressed multicast|bytes    packets errs drop fifo colls carrier compressed
    lo:     200       4    0    0    0     0          0         0      200       4    0    0    0     0       0          0
  🥲:         0       0    0    0    0     0          0         0        0       0    0    0    0     0       0          0
  toto:       0       0    0    0    0     0          0         0        0       0    0    0    0     0       0          0
```

Comment (mspncp) [reply scapy/arch/linux.py:237]: That convinces me :rofl:

Comment (mspncp) [reply scapy/arch/linux.py:237]: But it's still utf8, which is the default for decoding text files on linux. I wouldn't get an exception, would I?

Comment (gpotter2) [reply scapy/arch/linux.py:237]: Okay so you're actually right, but just for the sake of it let's push it to a theoretical madness. Here's the é character in latin-1 (e.g. default on many Windows devices)
```
user@host:~$ sudo ip link add � type veth peer name a
user@host:~$ # yeah that actually worked
user@host:~$ cat /proc/net/dev
Inter-|   Receive                                                |  Transmit
 face |bytes    packets errs drop fifo frame compressed multicast|bytes    packets errs drop fifo colls carrier compressed
    lo:     200       4    0    0    0     0          0         0      200       4    0    0    0     0       0          0
 bond0:       0       0    0    0    0     0          0         0        0       0    0    0    0     0       0          0
dummy0:       0       0    0    0    0     0          0         0        0       0    0    0    0     0       0          0
 tunl0:       0       0    0    0    0     0          0         0        0       0    0    0    0     0       0          0
  sit0:       0       0    0    0    0     0          0         0        0       0    0    0    0     0       0          0
  eth0: 427563127  265726    0    0    0     0          0      9435  4244917   55919    0    0    0     0       0          0
     a:       0       0    0    0    0     0          0         0        0       0    0    0    0     0       0          0
   �:       0       0    0    0    0     0          0         0        0       0    0    0    0     0       0          0
```

Comment (gpotter2) [reply scapy/arch/linux.py:237]: Obviously this is ridiculous and you're right, but let's keep it anyways :p

Comment (mspncp) [reply scapy/arch/linux.py:237]: Sure, the `errors='replace'` catch is useful without any doubt.

**Comment by gpotter2** [comment]
@p-l- @guedou Hi ! This is ready to merge, if one of you could have a look

File: scapy/arch/linux.py [hunk 3]
```diff
@@ -131,21 +132,12 @@ def _get_if_list():
     Function to read the interfaces from /proc/net/dev
     """
     try:
-        f = open("/proc/net/dev", "rb")
+        with open("/proc/net/dev", "r") as f:
```
Comment (mspncp) [inline scapy/arch/linux.py:135]:
```suggestion
        with open("/proc/net/dev", "r", errors='replace') as f:
```

File: scapy/arch/linux.py [hunk 4]
```diff
@@ -316,31 +329,28 @@ def in6_getifaddr():
     This is the list of all addresses of all interfaces available on
     the system.
     """
-    ret = []  # type: List[Tuple[str, int, str]]
     try:
-        fdesc = open("/proc/net/if_inet6", "rb")
+        with open("/proc/net/if_inet6", "r") as f:
```
Comment (mspncp) [inline scapy/arch/linux.py:333]:
```suggestion
        with open("/proc/net/if_inet6", "r", errors='replace') as f:
```

File: scapy/arch/linux.py [hunk 5]
```diff
@@ -351,43 +361,44 @@ def read_routes6():
...
-        cset = []  # candidate set (possible source addresses)
-        if dev == conf.loopback_name:
-            if d == '::':
-                continue
-            cset = ['::1']
-        else:
-            devaddrs = (x for x in lifaddr if x[2] == dev)
-            cset = scapy.utils6.construct_source_candidate_set(d, dp, devaddrs)
+    try:
+        with open("/proc/net/ipv6_route", "r") as f:
```
Comment (mspncp) [inline scapy/arch/linux.py:372]:
```suggestion
        with open("/proc/net/ipv6_route", "r", errors='replace') as f:
```

**Review by mspncp — COMMENTED** [review]
@gpotter2 since you introduced me to the `errors='replace'` parameter: wouldn't it be better to add it consistently to all open calls in text mode?

File: scapy/arch/linux.py [hunk 6]
```diff
@@ -238,68 +230,89 @@ def get_alias_address(iface_name,  # type: str
...
     # type: () -> List[Tuple[int, int, str, str, str, int]]
+    """
+    Read routes from /proc/net/route
+    """
     try:
-        f = open("/proc/net/route", "rb")
+        with open("/proc/net/route", "r", errors='replace') as f:
+            routes = []
+            # Loopback route
+            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
```
Comment (p-l-) [inline scapy/arch/linux.py:240]:
```suggestion
        with (
            open("/proc/net/route", "r", errors="replace") as f,
            socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s,
        ):
            routes = []
            # Loopback route
```

File: scapy/arch/linux.py [hunk 7]
```diff
@@ -238,68 +230,89 @@ def get_alias_address(iface_name,  # type: str
...
+                gw = scapy.utils.inet_ntoa(struct.pack("I", int(gw, 16)))
+
+                route = (dst, msk, gw, iff, ifaddr, metric)
+                if ifaddr_int & msk != dst:
+                    tmp_route = get_alias_address(iff, dst, gw, metric)
+                    if tmp_route:
+                        route = tmp_route
+                routes.append(route)
+
+            s.close()
```
Comment (p-l-) [inline scapy/arch/linux.py:310]:
```suggestion
```

**Review by p-l- — CHANGES_REQUESTED** [review]
Using `with` should also work for `socket.socket()` objects!

**Comment by mspncp** [comment]
@p-l-  According to the [Python documentation](https://docs.python.org/3/reference/compound_stmts.html#the-with-statement), grouping parentheses are only allowed since Python 3.10:

> Changed in version 3.10: Support for using grouping parentheses to break the statement in multiple lines.

That's why the [ubuntu-latest 3.7](https://github.com/secdev/scapy/actions/runs/4046498140/jobs/6959360998) and [ubuntu-latest 3.9](https://github.com/secdev/scapy/actions/runs/4046498140/jobs/6959361199) jobs fail. I'll fix that. (Interestingly, the [ubuntu-latest 3.8](https://github.com/secdev/scapy/actions/runs/4046498140/jobs/6959361370) job doesn't fail, although it should according to the documentation.)

**Comment by p-l-** [comment]
Thanks @mspncp! (and thanks @gpotter2)

**Comment by gpotter2** [comment]
Thanks @p-l- !

**Pull request merged**